srcts/src/components/_shinyResizeObserver.ts

```diff
@@ -0,0 +1,152 @@
+   * or the card body is expanded full screen. It performs two primary tasks:
+   *
+   * 1. Dispatches a `resize` event on the window object. This is necessary to
+   *    ensure that Shiny outputs resize appropriately. In general, the window
+   *    resizing is throttled and the output update occurs when the transition
+   *    is complete.
+   * 2. If an output with a resize method on the output binding is detected, we
+   *    directly call the `.onResize()` method of the binding. This ensures that
+   *    htmlwidgets transition smoothly. In static mode, htmlwidgets does this
+   *    already.
+   *
+   * @note
+   * This resize observer also handles race conditions in some complex
+   * fill-based layouts with multiple outputs (e.g., plotly), where shiny
+   * initializes with the correct sizing, but in-between the 1st and last
+   * renderValue(), the size of the output containers can change, meaning every
+   * output but the 1st gets initialized with the wrong size during their
+   * renderValue(). Then, after the render phase, shiny won't know trigger a
+   * resize since all the widgets will return to their original size (and thus,
+   * Shiny thinks there isn't any resizing to do). The resize observer works
+   * around this by ensuring that the output is resized whenever its container
+   * size changes.
+   * @constructor
+   */
+  constructor() {
+    this.resizeObserverEntries = [];
+    this.resizeObserver = new ResizeObserver((entries) => {
+      const resizeEvent = new Event("resize");
+      window.dispatchEvent(resizeEvent);
+
+      // the rest of this callback is only relevant in Shiny apps
+      if (!window.Shiny) return;
+
+      const resized = [] as HTMLElement[];
+
+      for (const entry of entries) {
+        if (!(entry.target instanceof HTMLElement)) continue;
+        if (!entry.target.querySelector(".shiny-bound-output")) continue;
+
+        entry.target
+          .querySelectorAll<HTMLElement>(".shiny-bound-output")
+          .forEach((el) => {
+            if (resized.includes(el)) return;
+
+            const { binding, onResize } = $(el).data("shinyOutputBinding");
+            if (!binding || !binding.resize) return;
+
+            // if this output is owned by another observer, skip it
+            const owner = (el as any).shinyResizeObserver;
+            if (owner && owner !== this) return;
+            // mark this output as owned by this shinyResizeObserver instance
+            (el as any).shinyResizeObserver = this;
+
+            // trigger immediate resizing of outputs with a resize method
+            onResize(el);
+            // only once per output and resize event
+            resized.push(el);
+          });
+
+        // set plot images to 100% width temporarily during the transition
+        entry.target
+          .querySelectorAll<HTMLImageElement>(
+            '.shiny-plot-output img:not([width="100%"])'
```

Comment:
```suggestion
            '.shiny-plot-output > img'
```

Reply:
Actually, now that I think of it, what if we did something like this in the `.forEach()` above (just before calling `onResize(el)` ?

```js
  if (el.classList.contains('shiny-plot-output')) {
    // should probably also check that it's a <img> tag
    el.children[0].setAttribute("width", "100%")
  }
```

Reply:
The reason for including `:not([width="100%"])` was to avoid doing extra work. We only need to make the update once, but the resize observer will trigger many times during the transition. The idea is to change to `width = "100%"` once when the first resize event occurs; after that we don't need to interact with the element.

I'm going to take another look at this, in particular around nested sidebars – e.g. page-level sidebar with internal sidebars.

Reply:
I update the test case to include a shared sidebar and local sidebars. I also did some benchmarking and I think our use of `querySelectorAll()` is a robust and fast-enough approach. We can actually speed it up a little bit by using `querySelector()` to first test if we have any plots before calling `querySelectorAll()`, so I'm going to apply that change.

Reply:
I moved the plot output width adjustment into the `.shiny-bound-output` visitor in cbe9cd8cbe9a7f0b80c2781ffe7a817bcf5e0639.